SDWebImage/SDWebImageCodersManager.m

```diff
@@ -0,0 +1,147 @@
+/*
+ * This file is part of the SDWebImage package.
+ * (c) Olivier Poitrey <rs@dailymotion.com>
+ * (c) james <https://github.com/mystcolor>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+#import "SDWebImageCodersManager.h"
+#import "SDWebImageImageIOCoder.h"
+#import "SDWebImageGIFCoder.h"
+#ifdef SD_WEBP
+#import "SDWebImageWebPCoder.h"
+#endif
+
+@interface SDWebImageCodersManager ()
+
+@property (nonatomic, strong, nonnull) NSMutableArray<SDWebImageCoder>* mutableCoders;
+@property (SDDispatchQueueSetterSementics, nonatomic, nullable) dispatch_queue_t mutableCodersAccessQueue;
+
+@end
+
+@implementation SDWebImageCodersManager
+
++ (nonnull instancetype)sharedInstance {
+    static dispatch_once_t once;
+    static id instance;
+    dispatch_once(&once, ^{
+        instance = [self new];
+    });
+    return instance;
+}
+
+- (instancetype)init {
+    if (self = [super init]) {
+        // initialize with default coders
+        _mutableCoders = [@[[SDWebImageImageIOCoder sharedCoder], [SDWebImageGIFCoder sharedCoder]] mutableCopy];
+#ifdef SD_WEBP
+        [_mutableCoders addObject:[SDWebImageWebPCoder sharedCoder]];
+#endif
+        _mutableCodersAccessQueue = dispatch_queue_create("com.hackemist.SDWebImageCodersManager", DISPATCH_QUEUE_CONCURRENT);
+    }
+    return self;
+}
+
+- (void)dealloc {
+    SDDispatchQueueRelease(_mutableCodersAccessQueue);
+}
+
+- (void)addCoder:(nonnull id<SDWebImageCoder>)coder {
+    if ([coder conformsToProtocol:@protocol(SDWebImageCoder)]) {
+        dispatch_barrier_sync(self.mutableCodersAccessQueue, ^{
+            [self.mutableCoders addObject:coder];
+        });
+    }
+}
+
+- (void)removeCoder:(nonnull id<SDWebImageCoder>)coder {
+    dispatch_barrier_sync(self.mutableCodersAccessQueue, ^{
+        [self.mutableCoders removeObject:coder];
+    });
+}
+
+- (NSArray<SDWebImageCoder> *)coders {
+    __block NSArray<SDWebImageCoder> *sortedCoders = nil;
+    dispatch_sync(self.mutableCodersAccessQueue, ^{
+        sortedCoders = (NSArray<SDWebImageCoder> *)[[[self.mutableCoders copy] reverseObjectEnumerator] allObjects];
```

Comment:
You're right. To avoid that user remove all the coders built-in. We need a nil check here and return the default coders array.

Reply:
I don't agree :) if the user wants to remove all the coders, so be it. Not our problem :)